src/fairseq2/datasets/utils.py

```diff
@@ -6,27 +6,40 @@
 
 import logging
 from logging import Logger
+from typing import List, Tuple, TypeVar
 
 import torch
 
 from fairseq2.gang import Gang
+from fairseq2.models import Batch
 
+BatchT = TypeVar("BatchT", bound=Batch)
 
-def _all_eod(eod: bool, gang: Gang, logger: Logger) -> bool:
-    """Return ``True`` if all processes in ``gang`` have reached end of data."""
-    if gang.size == 1:
-        return eod
-
-    eods = torch.empty((gang.size,), device=gang.device, dtype=torch.bool)
-
-    gang.all_gather(eods, torch.tensor(eod, device=gang.device))
 
-    if eods.any():
-        if logger.isEnabledFor(logging.DEBUG) and not eods.all():
-            ranks = ", ".join(str(r) for r in eods.nonzero().squeeze(1).tolist())
+def _all_batch_sizes(
+    batches: List[BatchT], gang: Gang, logger: Logger
+) -> Tuple[int, bool]:
+    """Return the sum of the batch sizes of all elements from all ranks."""
 
-            logger.debug(f"End of data reached at rank(s) {ranks}.")
-
-        return True
-
-    return False
+    total_bsz = 0
+    if gang.size == 1:
+        for batch in batches:
+            total_bsz += batch.batch_size
+        return total_bsz, total_bsz == 0
+
+    eod = False
+    for batch in batches:
+        batch_sizes = torch.empty((gang.size,), device=gang.device, dtype=torch.int64)
+        gang.all_gather(batch_sizes, torch.tensor(batch.batch_size, device=gang.device))
```

Comment:
This causes to call `all_gather()` `num_accumulate` times which is not efficient. Ideally you should call it just one with `total_bsz` of each rank.